### ISSUE TITLE
Feature: support java records in addition to java beans as config objects

### ISSUE THREAD

**Issue by pjroth** [lightbend/config#769]
`ConfigBeanFactory.create` supports the ability to create a typed object with a single method call which is very nice.  I am using java 17 and would like to use a `record` instead of a java bean to take advantage of the much reduced boilerplate possible when using java records.  

My proposal is something along the lines of the following:
```
record MyConfig(String uri, Integer port);

var map = Map.of(
                "uri", "localhost",
                "port", "1234");
var fullConfig = ConfigFactory.parseMap(map)
var myConfig = ConfigRecordFactory.create(fullConfig, MyConfig.class);

assertThat(myConfig.uri(), equalTo("localhost"));
assertThat(myConfig.port(), equalTo("1234"));
```

Thanks for this great library!

**Comment by tangjiafu** [comment]
I alse need this feature.

**Comment by mkurz** [comment]
I think you should provide a pull request if you want to see this happen.

**Comment by CharNikita** [comment]
As far as I can see, the project is not in active development, however, this feature would be very useful and convenient. Also, I see that the current open pull requests are not accepted. I can take on the implementation of this feature, but I would not like to waste time if the pull request will be rejected in any case.

@havocp You are the most active contributor to the library, can you please comment on whether I can take on the implementation?

**Comment by samvel-aivazian** [comment]
We definitely need this feature! Looking forward to seeing updates about it's completion!

**Comment by dkichler** [comment]
I think the main blocker for this feature in this library is it's continued support of Java 8, whereas records only came is as of Java 14 (as a preview).  Unless library authors intend to cut a branch/line that will drop support for Java 8 and move to 14 as the minimum, this feature will need to live elsewhere.  Since we have this feature in a project we use, I'm happy to open source it.  @havocp is there any intention of dropping Java 8 support on a distinct version line in the future?  I would be happy to contribute there if so, otherwise I can publish as as independent extension.

**Comment by ihostage** [comment]
Hi, @dkichler! It doesn't look like a blocker if use JEP 238: Multi-Release JAR Files and the related plugin https://github.com/sbt/sbt-multi-release-jar

**Comment by dkichler** [comment]
Looks like the core modules have dropped support for Java 8 as well:  https://doc.akka.io/libraries/akka-dependencies/current/java-versions.html

**Comment by dkichler** [comment]
@ihostage It appears https://github.com/sbt/sbt-multi-release-jar only has support for up to JDK11, with no clear path on how it will support a variety of newer versions.  I will go ahead with a release of the extension with minimum support of JDK 17 (the first LTS including records) for now and keep an eye out for a path to integration into the core libs going forward.

**Comment by ihostage** [comment]
@dkichler Yep, looks like for JDK 17 need to wait of this https://github.com/sbt/sbt-multi-release-jar/issues/24
